bitnami/redis/templates/configmap-scripts.yaml

```diff
@@ -208,19 +219,39 @@ data:
+      REDIS_MASTER_HOST="$(hostname -i)"
+      REDIS_MASTER_PORT_NUMBER="{{ .Values.redisPort }}"
     else
       if is_boolean_yes "$REDIS_SENTINEL_TLS_ENABLED"; then
-        sentinel_info_command="redis-cli -a $REDIS_PASSWORD -h {{ template "redis.fullname" . }}-headless.{{ .Release.Namespace }}.svc.{{ .Values.clusterDomain }} -p {{ .Values.sentinel.port }} --tls --cert ${REDIS_SENTINEL_TLS_CERT_FILE} --key ${REDIS_SENTINEL_TLS_KEY_FILE} --cacert ${REDIS_SENTINEL_TLS_CA_FILE} sentinel get-master-addr-by-name {{ .Values.sentinel.masterSet }}"
+        sentinel_info_command="redis-cli {{- if .Values.usePassword }} -a $REDIS_PASSWORD {{- end }} -h $HEADLESS_SERVICE -p {{ .Values.sentinel.port }} --tls --cert ${REDIS_SENTINEL_TLS_CERT_FILE} --key ${REDIS_SENTINEL_TLS_KEY_FILE} --cacert ${REDIS_SENTINEL_TLS_CA_FILE} sentinel get-master-addr-by-name {{ .Values.sentinel.masterSet }}"
       else
-        sentinel_info_command="redis-cli {{- if .Values.usePassword }} -a $REDIS_PASSWORD {{- end }} -h {{ template "redis.fullname" . }}-headless.{{ .Release.Namespace }}.svc.{{ .Values.clusterDomain }} -p {{ .Values.sentinel.port }} sentinel get-master-addr-by-name {{ .Values.sentinel.masterSet }}"
+        sentinel_info_command="redis-cli {{- if .Values.usePassword }} -a $REDIS_PASSWORD {{- end }} -h $HEADLESS_SERVICE -p {{ .Values.sentinel.port }} sentinel get-master-addr-by-name {{ .Values.sentinel.masterSet }}"
       fi
       REDIS_SENTINEL_INFO=($($sentinel_info_command))
       REDIS_MASTER_HOST=${REDIS_SENTINEL_INFO[0]}
       REDIS_MASTER_PORT_NUMBER=${REDIS_SENTINEL_INFO[1]}
-
-      sentinel_conf_set "sentinel monitor" "{{ .Values.sentinel.masterSet }} "$REDIS_MASTER_HOST" "$REDIS_MASTER_PORT_NUMBER" {{ .Values.sentinel.quorum }}"
     fi
+    sentinel_conf_set "sentinel monitor" "{{ .Values.sentinel.masterSet }} "$REDIS_MASTER_HOST" "$REDIS_MASTER_PORT_NUMBER" {{ .Values.sentinel.quorum }}"
+
+    add_replica() {
+      if [[ "$1" != "$REDIS_MASTER_HOST" ]]; then
+        sentinel_conf_add "sentinel known-replica {{ .Values.sentinel.masterSet }} $1 {{ .Values.redisPort }}"
+      fi
+    }
+
+    # remove generated known sentinels and replicas
+    tmp="$(sed -e '/^sentinel known-/d' -e '/^$/d' /opt/bitnami/redis-sentinel/etc/sentinel.conf)"
+    echo "$tmp" > /opt/bitnami/redis-sentinel/etc/sentinel.conf
+
+    for node in $(seq 0 {{ .Values.cluster.slaveCount }}); do
+      NAME="{{ template "redis.fullname" . }}-node-$node"
+      IP="$(getent hosts "$NAME.$HEADLESS_SERVICE" | awk ' {print $1 }')"
+      if [[ "$NAME" != "$HOSTNAME" && -n "$IP" ]]; then
+        sentinel_conf_add "sentinel known-sentinel {{ .Values.sentinel.masterSet }} $IP {{ .Values.sentinel.port }} $(host_id $NAME)"
+        add_replica "$IP"
+      fi
+    done
+    add_replica "$(hostname -i)"
```

Comment:
Could you explain this?

Reply:
TL;DR: in case the master is (temporarily) missing, the vanilla sentinels/redis system is not able to survive the changing IP addresses of sentinels/replicas (pods). This code attempts to seed the sentinel with everything that we can deduce from DNS. 

Background: I did try to kill half of the nodes in cluster (deleted the pods either directly or via evicting them from kubernetes nodes). I guess the problem was that when I (among the other pods) killed the acting master and there was not enough nodes to elect a new one and so the new/respawned pods could not connect to master (or node 0) to get information about the remaining nodes of the cluster. Ultimately I arrived at partitoned configurations (say, 2 surviving nodes knew about each other and 2 new ones knew each other but no communication between partitions). So I figured that I should supply information about existing nodes as deducted from DNS.

What the code does:

Function add_replica declares a known replica unless the IP address belongs to the declared/acting master. Without that condition, the sentinels would say to the master that it should be replica of itself (instead of replica of no one) and that will effectively disable that node. 

For-cycle iterates over all the nodes and if a) the name resolves (node is alive) and b) it is not the node itself, it adds the sentinel as known sentinel. It conditionally also adds replica as known replica. 

It finally also adds the node itself as replica (unless it is the master). The last bit is important for the sentinel(s) to be also able to reach that replica in case the master was down when the replica was started.

Reply:
Oh, .. and for that to work, one has to use static ids. I will wrap it into relevant if.

Reply:
Actually, it did work with non-static ids as well but logged that the ids were wrong.

Reply:
Thanks so much for the detailed explanation!! I don't see any caveats.
Anyway, I'd like @rafariossaa and @miguelaeh to review it too